### PR TITLE
Dullahan Partial Refactor: They Work Again Edition

### LINKED ISSUE #32236: Dullahans mouth is in their bodies
They can eat using their bodies
They can drink using their bodies
They can play russian roulette- ok you got it.

### LINKED ISSUE #41277: Dullahans dont die when they whisper in hardcrit
Issue reported from  Round ID: 96428 (/tg/Station Bagil [ENGLISH] [US-EAST] [100% FREE LAG])

Reporting client version: 512

very bad if there's a murder attempt and they're dragged away but someone finds their head, if they still have a bit of health left they can potentially give a very detailed explanation of what the murderer did

### LINKED ISSUE #41529: If you monkeyize a dullahan, the old head still exists
1. Monkeyize a dullahan
2. The monkey will have its head in its hand
3. The old head will still exist
4. Delete the old head
5. A lot of gibs will spawn in the location where the monkeyization happened

### LINKED ISSUE #54685: Unable to examine anything as a dullahan head, additionally dullahans can't use headsets
## Reproduction:
Step 1: Spawn as a dullahan, you'll have two heads (?)
Step 2: Toggle the perspective button a few times, otherwise you won't have the dullahan nerf at all and can dispose of the heads
Step 3: Using your primary head (the other head doesn't do anything) try to shift click on something to examine it, nothing will happen. If you use your blind body to examine stuff, it will fail because you're technically blind.

### LINKED ISSUE #62411: Dullahans' heads have different names than their bodies
Reporting client version: 514.1568

## Round ID:
[172065](https://scrubby.melonmesa.com/round/172065)

## Testmerges:
- [Vampire and Dullahan can be config enabled](https://www.github.com/tgstation/tgstation/pull/62357)
- [[DNM] [TM Candidate] Infuses certain vents with an aroma of garlic, repelling bat form vampires.](https://www.github.com/tgstation/tgstation/pull/61816)
- [Fixes issue with VALUE_MODE_FLAG config entries trying to parse values from keys.](https://www.github.com/tgstation/tgstation/pull/62359)

## Reproduction:
Be a Dullahan (player name was **Edgarius The Third**).
Examine your head. 
See head has a different name (head name was **Cruz Shick's head**)

### LINKED ISSUE #62441: Dullahans hear around the body instead of around the head
apologies I didn't get the round id.

Move your body and head to different locations. Try to talk to someone with the head and realise you cannot hear them. Meanwhile someone talks to your body and you hear them fine. But your ears are in your head! So what the hell.

### PR DESCRIPTION
**Pull request by GoldenAlpharex**
## About The Pull Request
So, a few months ago I was like "hmm there's something weird going on with party pods...", which got me looking into important_recursive_hearers or something like that. I spoke about it in the coding channel and Kyler actually fixed it before I did. But I also caught a similar glitch with Dullahans, so I decided to investigate...

*Two months later...*

I present to you a partial unfuckening of the Dullahans, in that I made them fully functional once again:
 - They only hear speech through their head (not sounds, sadly, someone else would have to tell me how to do that because I otherwise really wouldn't know how to do it in a sane way), they speak through their head, runechat-included.
 - When you spawn a Dullahan, you're set to look through the Dullahan's eyes (so from their head), and that doesn't reset when you log off and back in, or admin-ghost and come back in your body.
 - When you're looking through your head, your view will no longer be reset to your body upon entering a locker, which is nice to avoid not being blind while looking through your body.
 - Dullahan heads no longer look completely lifeless and without organs. They have eyes that don't look dead and that even match the player's intended eye color.
 - Dullahan can now properly examine things from their head, which was intended and 100% not functional.
 - Dullahan heads now speak with the proper name of their owner, instead of having a random name attached to it at round-start.
 - Dullahan heads are also now properly named too.
 - Dullahans can now properly whisper, sing and do all these funny things that they were unable to do before.
 - Dullahan whispers will now properly respect the range of the whisper.
 - Dullahans can now succumb in hardcrit by whispering, as intended. This potentially fixes other species that worked similarly not being able to succumb, like abductors, although I didn't test if they normally could, I just know they absolutely will be able to now.
 - When switching from Dullahans to a different species, your old head will no longer stay behind.

I also added a proc for species to do some code when we get a ckey login in our mob, which could potentially be useful for other stuff in the future, but it was necessary here as the view is reliant on the client, which we want to ensure doesn't get weird view glitches like having their head's vision overlay while actually being centered on their body.

I also made it so `say()` now takes a `range` argument, which is `7` by default, just so things that aren't humans can also whisper and do all those kinds of things. Going with that, there's probably a few more things that will be able to be done better thanks to this, although I haven't tested every edge case with this, but I doubt it will make *much* of a difference in the future.

Closes #32236
Closes #41277
Closes #41529
Closes #54685
Closes #62411
Closes #62441

There's a few issues I wasn't able to fix during this, or that I found out while making this PR, such as dullahans not having hair when they go back from being Monkeyfied, or the fact that skillchips fail to implant in Dullahans (which, to be fair, kinda makes sense most of the time, a possible fix would be a work-around when they're forced to take an implant, like when getting equiped with an outfit that has a skillchip, but not letting it happen otherwise, unless they're holding their brain in the machine or something, maybe), but I didn't deem them nearly important enough to make the species unplayable (like it currently is), and it's barely a minor annoyance at this point.

And, no, I don't know what happened to my branch, maybe I had it for a long time and just updated it or whatever, oh well.

## Why It's Good For The Game

That way, Dullahans are not a runtime-inducing buggy mess whenever Halloween happens, and they can *actually* be used.

I was just honestly ashamed to have something so utterly broken in our code, so I fixed it out of spite.

## Changelog

Lemon told me he made longer changelogs in the past so blame this on him if you think it's too long.

:cl: GoldenAlpharex
refactor: Refactored Dullahans in many areas to fix many, many of their bugs and glitches and ensure an actually functional experience playing as one, while also improving as much code as I could along the way.
fix: Dullahans only hear speech through their head (not sounds, sadly, someone else would have to tell me how to do that because I otherwise really wouldn't know how to do it in a sane way), they speak through their head, runechat-included.
fix: When you spawn a Dullahan, you're set to look through the Dullahan's eyes (so from their head), and that doesn't reset when you log off and back in, or admin-ghost and come back in your body.
fix: When you're looking through your head, your view will no longer be reset to your body upon entering a locker, which is nice to avoid not being blind while looking through your body.
fix: Dullahan heads no longer look completely lifeless and without organs. They have eyes that don't look dead and that even match the player's intended eye color.
fix: Dullahan can now properly examine things from their head's view.
fix: Dullahan heads now speak with the proper name of their owner, instead of having a random name attached to it at round-start.
fix: Dullahan heads are also now properly named too.
fix: Dullahans can now properly whisper, sing and do all these funny things that they were unable to do before.
fix; Dullahan whispers will now properly respect the range of the whisper.
fix: Dullahans can now succumb in hardcrit by whispering, as intended. This potentially fixes other species that worked similarly not being able to succumb, like abductors.
fix: When switching from Dullahans to a different species, your old head will no longer stay behind.
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
link all the issues this fixes please

**Comment by SuperNovaa41** [comment]
https://github.com/tgstation/tgstation/issues?q=is%3Aissue+is%3Aopen+dullahan+
i think these are all the current dullahan issues (hopefully this fixes all of them :))))

File: code/modules/mob/living/carbon/human/species_types/dullahan.dm [hunk 1]
```diff
@@ -20,57 +20,89 @@
...
+		human.gib()
+		return
+
+	if(my_head.loc.name != human.real_name && istype(my_head.loc, /obj/item/bodypart/head))
+		var/obj/item/bodypart/head/detached_head = my_head.loc
+		detached_head.real_name = human.real_name
+		detached_head.name = human.real_name
+		detached_head.brain.name = "[human.name]'s brain"
+
+	var/obj/item/bodypart/head/head2 = human.get_bodypart(BODY_ZONE_HEAD)
```
Comment (MrMelbert) [inline code/modules/mob/living/carbon/human/species_types/dullahan.dm:77]: Can you change this var name while you're here? `head2` tells nothing

**Comment by tralezab** [comment]
i'm just gonna send it, if dullahans brokey after this any more than they brokey now you will be dragged back kicking and screaming into dullahan.dm

**Comment by tralezab** [comment]
seems impossible to break dullahans any more though

**Pull request merged**